pjnat64/src/pjnat64/pj-nat64.c

```diff
@@ -0,0 +1,1409 @@
+/**
+ * Use algorithmic map
+ *
+ * @see https://datatracker.ietf.org/doc/html/rfc6052#section-2.1
+ */
+#ifndef USE_RFC6052
+# define USE_RFC6052 1
+#endif
+
+/* Interface */
+
+static void patch_sdp_attr(pj_pool_t *pool, pjmedia_sdp_media *media, pjmedia_sdp_attr *attr, pj_str_t addr_type, pj_str_t addr);
+static void patch_sdp_addr(pj_pool_t *pool, pjmedia_sdp_session *sdp, pj_str_t addr_type, pj_str_t addr);
+
+static void patch_msg_sdp_body_old(pj_pool_t *pool, pjsip_msg *msg, pjmedia_sdp_session *sdp);
+static void patch_msg_sdp_body(pj_pool_t *pool, pjsip_msg *msg, pjmedia_sdp_session *sdp);
+
+static void modern_sip_msg_dump(const char *title, pjsip_msg *msg);
+static void modern_sip_msg_sdp_dump(const char *title, pjsip_msg *msg);
+static void modern_sdp_dump(const char *title, pjmedia_sdp_session *sdp);
+
+static pjmedia_sdp_session *modern_sip_sdp_parse(pj_pool_t *pool, pjsip_msg *msg);
+
+static void modern_update_rx_data(pjsip_rx_data *rdata, pjmedia_sdp_session *sdp);
+static void modern_update_tx_data(pjsip_tx_data *tdata, pjsip_msg *msg);
+
+static void patch_sdp_ipv6_with_ipv4(pjsip_tx_data *tdata, pjsip_msg *msg);
+static void patch_sdp_ipv4_with_ipv6(pjsip_rx_data *rdata);
+
+/* Struct */
+
+struct addr {
+    pj_str_t in4_addr;  /* IPv4 address */
+    int      in4_port;  /* IPv4 port    */
+    pj_str_t in6_addr;  /* IPv6 address */
+    int      in6_port;  /* IPv6 port    */
+};
+
+/* Consts */
+
+static pj_caching_pool cp;
+static pj_bool_t mod_enable = PJ_TRUE;
+static pj_bool_t mod_debug = PJ_FALSE;
+static pj_pool_t *mod_pool = NULL;
+
+static struct addr hpbx;
+static struct addr acc;
+
+/* Implementation */
+
+/**
+ * Initialize address
+ *
+**/
+static void addr_init(struct addr *adr)
+{
+    /* Initialize IN4 addr */
+    adr->in4_addr.slen = 0;
+    adr->in4_addr.ptr = NULL;
+    adr->in4_port = -1;
+    /* Initialize IN6 addr */
+    adr->in6_addr.slen = 0;
+    adr->in6_addr.ptr = NULL;
+    adr->in6_port = -1;
+}
+
+static void addr_dump(const char *title, struct addr *adr)
+{
+    PJ_LOG(4, (THIS_FILE, "%s = [ IN4 = '%.*s':%d IN6 = '%.*s':%d ]",
+        title,
+        adr->in4_addr.slen, adr->in4_addr.ptr, adr->in4_port,
+        adr->in6_addr.slen, adr->in6_addr.ptr, adr->in6_port
+    ));
+}
+
+/**
+ * Modern SDP replace routine on SDP attribute
+ *
+**/
+static void patch_sdp_attr(pj_pool_t *pool, pjmedia_sdp_media *media, pjmedia_sdp_attr *attr, pj_str_t addr_type, pj_str_t addr)
+{
+    char rtcp_buf[128] = { 0 };
+    pj_str_t rtcp_name = pj_str("rtcp");
+    pjmedia_sdp_rtcp_attr ra;
+    pj_status_t s2;
+
+    if ((attr == NULL) || (attr->name.slen == 0)) {
+        return;
+    }
+
+    PJ_LOG(4, (THIS_FILE, "Process SDP attribute %.*s", attr->name.slen, attr->name.ptr));
+
+    if (strncmp(attr->name.ptr, rtcp_name.ptr, rtcp_name.slen) == 0) {
+        s2 = pjmedia_sdp_attr_get_rtcp(attr, &ra);
+        if (s2 != PJ_SUCCESS) {
+            PJ_LOG(1, (THIS_FILE, "Error: Patch SDP attribute 'rtcp'"));
```

Comment:
Can be a log, not an error log. Rtcp port not usually present in SDP.

Reply:
This is a trick to parse into `ra` struct `rtcp` attribute. Checking that we have an `rtcp` provide two lines earlier at `107`. In general, I think that there is probably a standard `rtcp` parsing function, but the documentation is the first thing I found.